packages/fx-core/src/plugins/resource/bot/clientFactory.ts

```diff
@@ -35,3 +41,36 @@ export function createWebSiteMgmtClient(
     throw new ClientCreationError(ClientNames.WEB_SITE_MGMT_CLIENT, e);
   }
 }
+
+export function createResourceProviderClient(
+  credentials: TokenCredentialsBase,
+  subscriptionId: string
+): Providers {
+  return new Providers(new ResourceManagementClientContext(credentials, subscriptionId));
+}
+
+export async function findResourceProvider(
+  client: Providers,
+  namespace: string
+): Promise<Provider | undefined> {
+  const provider = await client.get(namespace);
+  if (provider.registrationState?.trim() === "Registered") {
+    return provider;
+  }
+}
+
+export async function ensureResourceProvider(
+  client: Providers,
+  providerNamespaces: string[]
+): Promise<Provider[]> {
+  return Promise.all(
+    providerNamespaces.map(async (namespace) => {
+      const findedRP: Provider | undefined = await findResourceProvider(client, namespace);
+      if (!findedRP) {
+        return client.register(namespace);
```

Comment:
One idea comes to my mind. Just call client.register for once if the cost is identical.

Reply:
disscussed offline